rig-core/src/providers/anthropic/streaming.rs

```diff
@@ -0,0 +1,264 @@
+    pub stop_reason: Option<String>,
+    pub stop_sequence: Option<String>,
+}
+
+#[derive(Default)]
+struct ToolCallState {
+    name: String,
+    id: String,
+    input_json: String,
+}
+
+impl StreamingCompletionModel for CompletionModel {
+    async fn stream(
+        &self,
+        completion_request: CompletionRequest,
+    ) -> Result<
+        Pin<Box<dyn Stream<Item = Result<StreamingChoice, CompletionError>> + Send>>,
+        CompletionError,
+    > {
+        let prompt_with_context = completion_request.prompt_with_context();
+
+        let max_tokens = if let Some(tokens) = completion_request.max_tokens {
+            tokens
+        } else if let Some(tokens) = self.default_max_tokens {
+            tokens
+        } else {
+            return Err(CompletionError::RequestError(
+                "`max_tokens` must be set for Anthropic".into(),
+            ));
+        };
+
+        let mut request = json!({
+            "model": self.model,
+            "messages": completion_request
+                .chat_history
+                .into_iter()
+                .map(Message::from)
+                .chain(iter::once(Message {
+                    role: "user".to_owned(),
+                    content: prompt_with_context,
+                }))
+                .collect::<Vec<_>>(),
+            "max_tokens": max_tokens,
+            "system": completion_request.preamble.unwrap_or("".to_string()),
+            "stream": true,
+        });
+
+        if let Some(temperature) = completion_request.temperature {
+            merge_inplace(&mut request, json!({ "temperature": temperature }));
+        }
+
+        if !completion_request.tools.is_empty() {
+            merge_inplace(
+                &mut request,
+                json!({
+                    "tools": completion_request
+                        .tools
+                        .into_iter()
+                        .map(|tool| ToolDefinition {
+                            name: tool.name,
+                            description: Some(tool.description),
+                            input_schema: tool.parameters,
+                        })
+                        .collect::<Vec<_>>(),
+                    "tool_choice": ToolChoice::Auto,
+                }),
+            );
+        }
+
+        if let Some(ref params) = completion_request.additional_params {
+            merge_inplace(&mut request, params.clone())
+        }
+
+        let response = self
+            .client
+            .post("/v1/messages")
+            .json(&request)
+            .send()
+            .await?;
+
+        if !response.status().is_success() {
+            return Err(CompletionError::ProviderError(response.text().await?));
+        }
+
+        let stream = response.bytes_stream();
+        let (tx, rx) = mpsc::channel(100);
+
+        tokio::spawn(async move {
+            process_stream(stream, tx).await;
+        });
+
+        Ok(Box::pin(ReceiverStream::new(rx)))
```

Comment:
I don't think is needed to pin the stream and create a channel to consume it, normally you return the stream here and whenever is needed to be consumed you use `tokio::pin(stream)`.

Example, this function would be like this:
```rust
 async fn stream(
        &self,
        completion_request: CompletionRequest,
    ) -> impl TryStream<Ok=StreamingChoice, Err=CompletionError>> + Send {
. . .
    response.bytes_stream()
}
```

Then when I need to consume this stream:
```rust
#[tokio::main]
async fn main() -> Result<(), anyhow::Error> {
    let my_stream_model: StreamingCompletionModel  = ...;

    tokio::pin!(my_stream_model);

    while let Some(val) = my_stream_model.try_next().await? {
        print!(val);
    }
```

Reply:
@carlos-verdes there would be some additional issues:
* the building of completion request is async, so we would not be able to maintain the async interface and re-use the `Agent::completion` for generating a prompt + chat_history requests
* rather than returning the pin (current impl), we pin it afterwards to have the retval `await: <TryNext<'_, Pin<&mut impl TryStream<Ok = StreamingChoice, Error = CompletionError>>> as IntoFuture>::Output`

while I gave it a go (https://github.com/piotrostr/rig/commit/58c31dcd707700673043f378a200e6b57e69ba1c), going to sort the interfaces first, if you know exactly how to sort it please let me know, might be missing something here

Reply:
I sent full example with one of my libraries (it's code from a working project) where I return impl Stream on functions that actually are not async, you can simply wrap the side effect inside the stream.

More or less the idea is the effect call (for example a http request) is wrapped inside the Stream, something like:
- if request is ok, then emit one Stream element per response result
- if request returns error, then emit one Stream error event 

The only point where you need async/pin is when you consume the stream, this mean that all the logic around streams is fully lazy, there is non blocking until you actually need to consume the stream.